xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -4266,21 +4309,26 @@ Status IrEmitterUnnested::EmitIRForReduction(
 
 Status IrEmitterUnnested::EmitUnnestedReduction(
     mlir::lmhlo::FusionOp fusion, HloFusionAnalysis& fusion_analysis) {
+  HloComputation* fused_computation =
+      const_cast<HloComputation*>(fusion_analysis.fused_computation());
   auto* reduction_codegen_info = fusion_analysis.GetReductionCodegenInfo();
   // Set flag to false as Reduction has it's own custom logic of choosing a
   // block size.
   TF_ASSIGN_OR_RETURN(auto launch_dimensions,
                       fusion_analysis.GetLaunchDimensions(
                           /*use_experimental_block_size=*/false));
 
+  auto hlo_roots = GetFusionRoots(fused_computation);
   VLOG(3) << "Launch dimensions of "
           << mlir::mhlo::GetDebugNameFromLocation(fusion.getLoc()) << ": "
           << launch_dimensions.ToString();
   if (!reduction_codegen_info->IsRaceFree()) {
     absl::Span<HloInstruction* const> fusion_roots =
         fusion_analysis.fusion_roots();
     for (int i = 0; i < fusion_roots.size(); ++i) {
-      if (IsReductionFromOrToContiguousDimensions(*fusion_roots[i])) {
+      const HloInstruction* reduction_hero =
+          FindRealReductionHero(hlo_roots[i]);
+      if (reduction_hero->opcode() == HloOpcode::kReduce) {
```

Comment:
I would suggest another helper method HasRealReductionHero(), as you need this kind of check (calling FindRealReductionHero + verify it returns a reduce) several times.

Reply:
sounds good.